### PR TITLE
Support GEOS Cube-Sphere Grids

### LINKED ISSUE #801: Support GEOS Cube Sphere Grids
https://github.com/GEOS-ESM/MAPL/discussions/2844

### PR DESCRIPTION
**Pull request by philipc2**
Closes #801 

## Overview

* Add support for reading and encoding the [GEOS Cube-Sphere](https://gmao.gsfc.nasa.gov/gmaoftp/ops/GEOSIT_sample/doc/CS_Description_c180_v1.pdf) format in the UGRID conventions

### CODE REVIEW

**Comment by philipc2** [comment]
Plot of a ~900 element grid (CS12)

image

**Comment by philipc2** [comment]
@mathomp4

Let me know if you have any feedback!

**Comment by mathomp4** [comment]
> @mathomp4
> 
> Let me know if you have any feedback!

Well, I think I need to learn more Python but I decided to try on a stretched grid file I finally got. I first tried plotting the grid but apparently this:
```
<uxarray.Grid>
Original Grid Type: GEOS-CS
Grid Dimensions:
  * n_node: 440646
  * n_face: 437400
  * n_max_face_nodes: 4
  * n_nodes_per_face: (437400,)
Grid Coordinates (Spherical):
  * node_lon: (440646,)
  * node_lat: (440646,)
  * face_lon: (437400,)
  * face_lat: (437400,)
Grid Coordinates (Cartesian):
Grid Connectivity Variables:
  * face_node_connectivity: (437400, 4)
Grid Descriptor Variables:
```
was a bit too much for my laptop. I waited a few minutes and in the end it was just solid blue.

But following some of your examples, I did get out:

c270_plot

which is nice! I then tried a `plot.polygons()` and my computer hated me again. 😄 

I think what this demonstrates is that I need to convince my boss to help make a much coarser stretch grid...or learn more UXarray

**Comment by philipc2** [comment]
Glad you were able to get it working! Surprised that the Polygon plot took that long. Typically grids of that size should take no longer than a minute maximum for the first run.

**Comment by mathomp4** [comment]
> Glad you were able to get it working! Surprised that the Polygon plot took that long. Typically grids of that size should take no longer than a minute maximum for the first run.

Hmm. Well, it was on an M3 MacBook Air...if that matters? 

Note: it's also possible I called things incorrectly. I was just sort of trying to echo the cookbook. 😄

File: uxarray/io/utils.py [hunk 1]
```diff
@@ -37,6 +37,8 @@ def _parse_grid_type(dataset):
         mesh_type = "MPAS"
     elif "maxNodePElement" in dataset.dims:
         mesh_type = "ESMF"
+    elif all(key in dataset.sizes for key in ["nf", "YCdim", "XCdim"]):
```
Comment (rajeeja) [inline uxarray/io/utils.py:40]: I think this check might not always work, as some meshes that have a uniform number of nodes per element, maxNodePElement might not be explicitly stated.

Also, maybe put a comment or ref for what these keys mean. 

```
nf: This is the number of faces in the grid. For a cubed-sphere grid, this is typically 6 (representing the 6 faces of a cube).

YCdim: This is the Y-coordinate dimension. It represents the number of grid cells in the Y direction on each face of the cubed-sphere.

XCdim: This is the X-coordinate dimension. It represents the number of grid cells in the X direction on each face of the cubed-sphere.
```

**Review by rajeeja — CHANGES_REQUESTED** [review]

**Comment by philipc2** [comment]
> > Glad you were able to get it working! Surprised that the Polygon plot took that long. Typically grids of that size should take no longer than a minute maximum for the first run.
> 
> Hmm. Well, it was on an M3 MacBook Air...if that matters?
> 
> Note: it's also possible I called things incorrectly. I was just sort of trying to echo the cookbook. 😄

Yeah, I've run similar sized meshes on a 32GB M1 and had reasonable performance. Mind sharing the file so I can try to recreate it?

File: uxarray/io/utils.py (hunk 1 above)
Comment (philipc2) [reply uxarray/io/utils.py:40]: Good suggestion!

Comment (philipc2) [reply uxarray/io/utils.py:40]: > I think this check might not always work, as some meshes that have a uniform number of nodes per element, maxNodePElement might not be explicitly stated.

The `maxNodePElement` is not part of this PR, but I agree there are some meshes that are stored in a 1D format as opposed to a 2D array.

**Comment by mathomp4** [comment]
> Yeah, I've run similar sized meshes on a 32GB M1 and had reasonable performance. Mind sharing the file so I can try to recreate it?

Yes...if I can figure out how to share it. Can you try:

https://gmao.gsfc.nasa.gov/gmaoftp/gmao_SIteam/test-fixapp-c270.geosgcm_prog_nat.20200415_0000z.nc4

I think that works. If you get it, let me know. It's about 500 MB because I didn't really clean it up

**Comment by philipc2** [comment]
```Python
import uxarray as ux
import geoviews.feature as gf
import cartopy.crs as ccrs

features = gf.coastline(
    projection=ccrs.PlateCarree(), line_width=1, scale="50m"
) * gf.states(projection=ccrs.PlateCarree(), line_width=1, scale="50m")

file_path = r"C:\Users\chmie\Downloads\test-fixapp-c270.geosgcm_prog_nat.20200415_0000z.nc4"
```

```Python
%%time
uxds = ux.open_dataset(file_path, file_path)
```

Took about 3s

```Python
%%time
uxds['T'][0][-1].plot.rasterize(method='polygon', exclude_antimeridian=True) * features
```

Took about 3s as well.

image


```Python
uxds['RH'][0][0].subset.bounding_circle((-90, 25), 10).plot.rasterize(method='polygon', exclude_antimeridian=True, pixel_ratio=4.0) * features
```

image

**Comment by philipc2** [comment]
I'd like to mention that the white spots are due to an upstream issue with Datashader, which will be fixed in the next release.

https://github.com/holoviz/datashader/pull/1329

**Comment by mathomp4** [comment]
Hmm. When I first tried your first block of code, I got:
```
ValueError: No plotting extension is currently loaded. Ensure you load an plotting extension with hv.extension or import it explicitly from holoviews.plotting before applying any options.
```

So, I added:
```
import holoviews as hv
hv.extension("bokeh")
```

And then the plots took about 2s. So...yeah, I must have *really* screwed things up in my testing to be ~100x slower. 😄

**Comment by mathomp4** [comment]
And I was wondering about the weird white spots, but I figured they were due to just...something.

Now, my laptop was not happy with:
```
uxgrid = ux.open_grid(file_path,file_path)

%%time
uxgrid.plot()
```
Actually, my browser sort of crashed before it could render...

**Comment by philipc2** [comment]
uxgrid = ux.open_grid(file_path,file_path)

%%time
uxgrid.plot()

> And I was wondering about the weird white spots, but I figured they were due to just...something.
> 
> Now, my laptop was not happy with:
> 
> ```
> uxgrid = ux.open_grid(file_path,file_path)
> 
> %%time
> uxgrid.plot()
> ```
> 
> Actually, my browser sort of crashed before it could render...

The issue here is that you're using `open_grid` instead of `open_dataset`. The default plotting for a `Grid` object is an edge plot of the mesh, which is indeed very slow for any mesh larger than a few thousand elements.

**Comment by philipc2** [comment]
> Ohhhh. So:
> 
> ```
> uxds = ux.open_dataset()
> uxds.uxgrid.plot()
> ```
> 
> would be faster? I'll try that tonight on my laptop!
> 
> (I'm learning!)

Plotting a data variable would be faster, since that performs a raster operation as opposed to a vector plot of the geometries.

Our grid (i.e. only mesh) plotting is all vector based, but our data variables can be plotted as rasters.

```Python
# defaults to raster plot
uxds['T'].plot()

# can specifically call a raster
uxds['T'].plot.rasterize(method='polygon')
```

File: uxarray/io/_geos.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import xarray as xr
+import numpy as np
+
+from uxarray.constants import INT_DTYPE
+from uxarray.conventions import ugrid
+
+
+def _read_geos_cs(in_ds: xr.Dataset):
```
Comment (aaronzedwick) [inline uxarray/io/_geos.py:8]: Does this grid format support cartesian?

Comment (philipc2) [reply uxarray/io/_geos.py:8]: I don't believe so. 

@mathomp4 Are the coordinates typically spherical or are they every provided as Cartesian?

Comment (mathomp4) [reply uxarray/io/_geos.py:8]: Umm. I'm going to mention @bena-nasa here. He's our grids guru. I mainly know: our native CS output looks like this when I ask for it 😄 

We can also put out lat-Lon output that is of a form everything can read. But the CS output is the weird one. (And I think our restarts might be a slightly different CS unstructured grid style...)

Comment (philipc2) [reply uxarray/io/_geos.py:8]: Resolving this comment for now, as we don't need Cartesian coordinates anyways. We can always add support for parsing them if it turns out they are supported by GEOS in the future.

**Review by aaronzedwick — APPROVED** [review]
Nice work on this!

**Comment by mathomp4** [comment]
> Nice work on this!

Doubly so from me and the GEOS community. Nice to have another package that can read our files without need to do all sort of gymnastics!

**Comment by mathomp4** [comment]
Woo! Thanks. I eagerly await the next release! 😄

**Comment by mathomp4** [comment]
> I'd like to mention that the white spots are due to an upstream issue with Datashader, which will be fixed in the next release.
> 
> [holoviz/datashader#1329](https://github.com/holoviz/datashader/pull/1329)

Can confirm, speckles are gone...though I am getting a lot of:
```
/Users/mathomp4/installed/Core/GEOSpyD/24.4.0-0_py3.11/2024-06-11/lib/python3.11/site-packages/xarray/namedarray/core.py:496: UserWarning: Duplicate dimension names present: dimensions {'ncontact'} appear more than once in dims=('nf', 'ncontact', 'ncontact'). We do not yet support duplicate dimension names, but we do allow initial construction of the object. We recommend you rename the dims immediately to become distinct, as most xarray functionality is likely to fail silently if you do not. To rename the dimensions you will need to set the ``.dims`` attribute of each variable, ``e.g. var.dims=('x0', 'x1')``.
  warnings.warn(
```
I can get a picture, so this is working, but is this something we should worry about in future?

**Pull request merged**